### PR TITLE
enh(openid): update configuration form to handle auth conditions

### PR DESCRIPTION
**Pull request by jeremyjaouen**
## Description

this PR intends to add new authentication_conditions parameters into open id configuration form.

This PR handle **only** the configuration form and not the authentication.

This PR target MON-14395 and should wait this branch will be merged on develop before going out from draft.

**Fixes** # MON-15004

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## Target serie

- [ ] 21.04.x
- [ ] 21.10.x
- [ ] 22.04.x
- [x] 22.10.x (master)

#### Community contributors & Centreon team

- [x] I have followed the **coding style guidelines** provided by Centreon
- [x] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

File: src/Core/Security/AccessGroup/Infrastructure/Repository/DbReadAccessGroupRepository.php [hunk 1]
```diff
@@ -218,20 +218,24 @@ public function findByIds(array $accessGroupIds): array
...
-        );
-        foreach ($queryBindValues as $bindKey => $accessGroupId) {
-            $statement->bindValue($bindKey, $accessGroupId, \PDO::PARAM_INT);
-        }
-        $statement->execute();
 
         $accessGroups = [];
-        while ($statement !== false && is_array($result = $statement->fetch(\PDO::FETCH_ASSOC))) {
-            $accessGroups[] = DbAccessGroupFactory::createFromRecord($result);
+        if (! empty($queryBindValues)) {
```
Comment (TamazC) [inline src/Core/Security/AccessGroup/Infrastructure/Repository/DbReadAccessGroupRepository.php]: you could use early return
if  (empty($queryBindValues)) {
........
}

File: src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php [hunk 2]
```diff
@@ -184,4 +180,20 @@ public static function authorizationRulesToArray(array $authorizationRules): arr
             ];
         }, $authorizationRules);
     }
+
+    /**
+     * @param AuthenticationConditions $authenticationConditions
+     * @return array<string,bool|string|string[]>
+     */
+    public static function authenticationConditionsToArray(AuthenticationConditions $authenticationConditions): array
```
Comment (TamazC) [inline src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php:188]: You have the same function into "DbWriteOpenIdConfigurationRepository.php" file.
Maybe you could put this method iside "AuthenticationConditions" class ?

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/AuthenticationConditions.php [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            }
+            if (empty($endpoint)) {
+                $mandatoryParameters[] = "endpoint";
+            }
+            if (empty($authorizedValues)) {
+                $mandatoryParameters[] = "authorized_values";
+            }
+            if (! empty($mandatoryParameters)) {
+                throw OpenIdConfigurationException::missingMandatoryParameters($mandatoryParameters);
+            }
```
Comment (TamazC) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/AuthenticationConditions.php]: I'd create a method for this kind of check like

```
    private function checkMandatoryParams(): void
    {
        if (!$this->isEnabled) {
            return;
        }

        $missingParams = [];
        if(empty($this->attributePath)) {
            $missingParams[] = 'attribute_path';
        }

        if (empty($this->endpoint)) {
            $missingParams[] = 'endpoint';
        }

        if (empty($this->authorizedValues)) {
            $missingParams[] = 'authorized_values';
        }

        throw OpenIdConfigurationException::missingMandatoryParameters($missingParams);
    }
```

File: src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php (hunk 2 above)
Comment (jeremyjaouen) [reply src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/FindOpenIdConfiguration/FindOpenIdConfigurationResponse.php:188]: agreed but not in the Model as it is not related to business logic but presentation, we'll discuss about it !

File: src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/UpdateOpenIdConfiguration/UpdateOpenIdConfiguration.php [hunk 4]
```diff
@@ -258,4 +262,33 @@ private function updateConfiguration(Configuration $configuration): void
...
+        if (! empty($authenticationConditionsParameters["trusted_client_addresses"])) {
+            $authenticationConditions->setTrustedClientAddresses(
+                $authenticationConditionsParameters["trusted_client_addresses"]
+            );
+        }
+        if (! empty($authenticationConditionsParameters["blacklist_client_addresses"])) {
+            $authenticationConditions->setBlacklistClientAddresses(
+                $authenticationConditionsParameters["blacklist_client_addresses"]
+            );
+        }
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Application/OpenId/UseCase/UpdateOpenIdConfiguration/UpdateOpenIdConfiguration.php]:
```suggestion
            $authenticationConditions->setTrustedClientAddresses(
                $authenticationConditionsParameters["trusted_client_addresses"]
            );
            $authenticationConditions->setBlacklistClientAddresses(
                $authenticationConditionsParameters["blacklist_client_addresses"]
            );
```

File: src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/AuthenticationConditions.php [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+        ) {
+            throw AssertionException::ipOrDomain(
+                $clientAddress,
+                'AuthenticationConditions::' . $fieldName
+            );
+        }
+    }
+
+    /**
+     * Undocumented function
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Domain/OpenId/Model/AuthenticationConditions.php]:
```suggestion
     * Undocumented function
```
add documentation

File: src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Repository/CustomConfigurationSchema.json [hunk 6]
```diff
@@ -1,103 +1,186 @@
...
+    "connection_scopes",
+    "login_claim",
+    "client_id",
+    "client_secret",
+    "authentication_type",
+    "verify_peer",
+    "auto_import",
+    "contact_template_id",
+    "email_bind_attribute",
+    "fullname_bind_attribute"
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Repository/CustomConfigurationSchema.json]:
```suggestion
    "fullname_bind_attribute",
    "authentication_conditions"
```

File: src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Repository/CustomConfigurationSchema.json [hunk 7]
```diff
@@ -1,103 +1,186 @@
...
+      "items": {
+        "type": "string"
+      }
+    },
+    "blacklist_client_addresses": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      }
+    },
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Infrastructure/OpenId/Repository/CustomConfigurationSchema.json]:
```suggestion
```

File: src/Core/Security/ProviderConfiguration/Infrastructure/Repository/DbReadConfigurationRepository.php [hunk 8]
```diff
@@ -253,4 +257,34 @@ public function findExcludedUsers(): array
...
+        if (! empty($authenticationConditionsRecord["trusted_client_addresses"])) {
+            $authenticationConditions->setTrustedClientAddresses(
+                $authenticationConditionsRecord["trusted_client_addresses"]
+            );
+        }
+        if (! empty($authenticationConditionsRecord["blacklist_client_addresses"])) {
+            $authenticationConditions->setBlacklistClientAddresses(
+                $authenticationConditionsRecord["blacklist_client_addresses"]
+            );
+        }
```
Comment (jeremyjaouen) [inline src/Core/Security/ProviderConfiguration/Infrastructure/Repository/DbReadConfigurationRepository.php]:
```suggestion
            $authenticationConditions->setTrustedClientAddresses(
                $authenticationConditionsRecord["trusted_client_addresses"]
            );
            $authenticationConditions->setBlacklistClientAddresses(
                $authenticationConditionsRecord["blacklist_client_addresses"]
            );
```

**Review by kduret — APPROVED** [review]

**Review by adr-mo — APPROVED** [review]

**Pull request merged**